vignettes/cookbook.Rmd

```diff
@@ -0,0 +1,221 @@
+```
+
+(the `broom` package provides a nice way of doing this sort of manipulation of these slightly opaque objects).  There are many ways of serialising this sort of data; we will do it in the default way supported by `jsonlite`, representing the object as an array of objects, each of which is key/value pairs for each row:
+
+```{r}
+jsonlite::toJSON(lm_coef, pretty = TRUE)
+```
+
+So we have our target function now:
+
+```{r}
+fit_lm <- function(data) {
+  data <- jsonlite::fromJSON(data)
+  fit <- lm(y ~ x, data)
+  lm_coef <- as.data.frame(summary(fit)$coefficients)
+  lm_coef <- cbind(name = rownames(lm_coef), lm_coef)
+  rownames(lm_coef) <- NULL
+  lm_coef
+}
+```
+
+Note that the target function must deserialise the json itself. This is so that arguments can be passed to `jsonlite::fromJSON` easily to control how deserialisation happens.  We may support automatic deserialisation later as an argument to `porcelain::porcelain_input_body_json`.
+
+The endpoint is not that much more involved than before though we have interesting inputs and outputs, with schemas required for both
+
+```{r}
+endpoint_lm <- porcelain::porcelain_endpoint$new(
+  "POST", "/lm", fit_lm,
+  porcelain::porcelain_input_body_json("data", "lmInputs", schema_root),
+  returning = porcelain::porcelain_returning_json("lmCoef", schema_root))
+```
+
+The input schema, `lmInputs.json` is
+
+```{r, echo = FALSE, results = "asis"}
+show_file(file.path(schema_root, "lmInputs.json"), "json")
+```
+
+while the response schema `lmCoef.json` is
+
+```{r, echo = FALSE, results = "asis"}
+show_file(file.path(schema_root, "lmCoef.json"), "json")
+```
+
+These are both fairly strict schemas using both `required` and `additionalProperties`. You might want to be more permissive, but we find that strictness here pays off later.
+
+```{r}
+api <- porcelain::porcelain$new(validate = TRUE)$handle(endpoint_lm)
+```
+
+To exercise the API endpoint we need to pass in our input JSON (not an R object).
+
+```{r}
+json <- jsonlite::toJSON(data)
+json
+```
+
+```{r}
+api$request("POST", "/lm", body = json, content_type = "application/json")
+```
+
+# POST endpoint, inputs as binary, returning binary
+
+(This example also shows off a few other features)
+
+Handling binary inputs and outputs is supported, provided that you can deal with them in R.  In this example we'll use R's serialisation format (rds; see `?serialize` and `?saveRDS`) as an example, but this approach would equally work with excel spreadsheets, zip files or any other non-text data that you work with.
+
+In this example we'll take some serialised R data and create a png plot as output.  We'll start by writing our target function:
+
+```{r}
+binary_plot <- function(data, width = 400, height = 400) {
+  data <- unserialize(data)
+  tmp <- tempfile(fileext = ".png")
+  on.exit(unlink(tmp))
+  png(tmp, width = width, height = height)
+  tryCatch(
+    plot(data),
+    finally = dev.off())
+  readBin(tmp, raw(), n = file.size(tmp))
```

Comment:
Could include how to add a header here too
```
porcelain::porcelain_add_headers(bytes, list("Content-Type" = "image/png")) 
```
or fine to leave it out if you want to keep the example simple

Reply:
Hmm, should we do this differently/elsewhere? Should `porcelain::porcelain_returning_binary` accept a content_type field (some of the inputs do I think). Might be worth thinking about this properly...

Reply:
https://mrc-ide.myjetbrains.com/youtrack/issue/RESIDE-261